### PR TITLE
[Fix #63] Avoid spurious output by using buffer redirection

### LINKED ISSUE #63: Spurious output when first opened in lumo
## Expected behavior

When first opening the repl and cursor is on code buffer, only `cljs.user =>` should appear.

## Actual behavior

The output of a command (for me it is arglists because the cursor was after fn symbol, but this varies) is printed in the `*inf-clojure*` buffer.

## Steps to reproduce the problem

`C-c C-z` with popup window enabled and move the cursor around in the code buffer.

or in `ielm`:

```
ELISP> (setq inf-clojure-repl-type nil)
nil
ELISP> (inf-clojure-arglist 'defn)
nil
```

### inf-clojure version information

*Latest master*

### Lein/Boot version

*Lumo latest master*

Reminder to self :sweat: :grimacing: :grinning:

### PR DESCRIPTION
**Pull request by arichiardi**
I will need to fix the arglists function as well, it does not work properly for edge cases, I already have some code ready, but I figured I should do another PR.

Also, the same approach should be used for completions, but completions require more work that is why I avoided putting it here.

-----------------

Before submitting the PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [X] The commits are consistent with our [contribution guidelines][1]
- [X] You've updated the changelog (if adding/changing user-visible functionality)

Thanks!

[1]: https://github.com/clojure-emacs/inf-clojure/blob/master/.github/CONTRIBUTING.md

### CODE REVIEW

**Comment by arichiardi** [comment]
The problem with completions is that you have to post-process them in different ways depending on the repl type. For instance lumo returns json, which emacs can read. So completions need a better handling.

File: inf-clojure.el [hunk 1]
```diff
@@ -884,22 +884,46 @@ safe to call only from inside `inf-clojure-arglist`."
...
-    (unwind-protect
-        (let ((eldoc-snippet (format (inf-clojure-arglists-form) fn)))
-          (inf-clojure--send-string proc eldoc-snippet)
-          (while (and (not (string-match inf-clojure-prompt kept))
-                      (accept-process-output proc 2)))
-          (setq eldoc (inf-clojure-match-arglists eldoc-snippet kept)))
-      (set-process-filter proc comint-filt))
-    eldoc))
+  (let ((eldoc-snippet (format (inf-clojure-arglists-form) fn)))
+    (inf-clojure-results-from-process (inf-clojure-proc) (concat eldoc-snippet "\n"))))
```
Comment (bbatsov) [inline inf-clojure.el]: Why don't you ensure the command ends with newline in `inf-clojure-results-from-process`?

File: inf-clojure.el [hunk 2]
```diff
@@ -1145,20 +1169,9 @@ to suppress the usage of the target buffer discovery logic."
...
-        ;; use the real comind-send-string in order to avoid stack overflows
-        (comint-send-string proc (concat form "\n"))
-      ;; yey, loads of procedural code again
-      (while (and (not is-matching)
-                  (not (string-match inf-clojure-prompt kept))
-                  (accept-process-output proc 2))
-        (setq is-matching (funcall match-p kept)))
-      (set-process-filter proc prev-filter))
-    is-matching))
+  (thread-last
```
Comment (bbatsov) [inline inf-clojure.el]: The use of `thread-last` here seems pointless to me.

File: inf-clojure.el (hunk 1 above)
Comment (arichiardi) [inline inf-clojure.el]: Yeah maybe that's better

File: inf-clojure.el (hunk 2 above)
Comment (arichiardi) [inline inf-clojure.el]: The command was so long that is why I used it, I can remove it

**Comment by arichiardi** [comment]
Fixed the issue, I will need another sweep for fixing the problems with arglists and completions once and for all. Also, we could remove all the `\n` at the end of the form now, but I haven't included it here.

**Pull request merged**